olive/passes/onnx/rtn_quantization.py

```diff
@@ -0,0 +1,349 @@
+        ir_model = model.load_ir_model()
+        ir_model.graph.opset_imports[MSFT_DOMAIN] = 1
+        self._quantize_model(ir_model, config.nodes_to_exclude, config.nodes_to_include)
+        return ir_model_to_olive_model(ir_model, output_model_path, config)
+
+    def _quantize_model(
+        self,
+        ir_model: ir.Model,
+        nodes_to_exclude: Optional[list[str]] = None,
+        nodes_to_include: Optional[list[str]] = None,
+    ):
+        nodes_to_exclude = nodes_to_exclude or []
+        nodes_to_include = nodes_to_include or []
+
+        ir_model.graph.sort()
+        for node in ir.traversal.RecursiveGraphIterator(ir_model.graph):
+            node_name = node.name
+
+            if node_name in nodes_to_exclude:
+                logger.debug("exclude to quantize %s as specified by nodes_to_exclude...", node_name)
+                continue
+
+            elif node.op_type in (str(OpType.MatMul), str(OpType.Gather)) and (
+                node_name in nodes_to_include or not nodes_to_include
+            ):
+                if (node.op_type == OpType.MatMul and not node.inputs[1].is_initializer()) or (
+                    node.op_type == OpType.Gather and not node.inputs[0].is_initializer()
+                ):
+                    continue
+
+                quantized_node = self.quantize(node)
+
+                if quantized_node.op_type == OpType.MatMulNBits or quantized_node.op_type == OpType.GatherBlockQuantized:
+                    for input_value in quantized_node.inputs:
+                        if input_value.is_initializer():
+                            node.graph.register_initializer(input_value)
+
+                    ir.convenience.replace_nodes_and_values(
+                        node.graph, node, [node], [quantized_node], node.outputs, quantized_node.outputs
+                    )
+            else:
+                logger.debug("skip to quantize %s ...", node_name)
+
+    def quantize(self, node: ir.Node) -> ir.Node:
+        """Quantize the weight of the target node and return the new nodes.
+
+        Target node:        QOperator node:
+        MatMul              MatMulNBits
+        Gather              GatherBlockQuantized
+        If the node is target node with fp32 or fp16 const weight, quantize the weight to int4 and
+        return the new nodes.
+        return the corresponding QOperator nodes.
+        """
+        logger.debug("Start quantizing %s ...", node.name)
+
+        if node.op_type == OpType.MatMul:
+            return self.quantize_matmul(node)
+        elif node.op_type == OpType.Gather:
+            if self.bits != 4:
+                logger.error("Gather only supports 4-bit quantization.")
+                return node
+            return self.quantize_gather(node)
+        else:
+            logger.error("Unsupported op %s for weight-only quantization.", node.op_type)
+            return node
+
+    def quantize_matmul(self, node: ir.Node) -> ir.Node:
```

Comment:
```suggestion
    def quantize_matmul(self, node: ir.Node) -> tuple[ir.Node, ir.Graph]:
```

Potentially also return the owning graph of the original weights, so that the new initializers can be added correctly